overlord/servicestate/quota_handlers.go

```diff
@@ -37,6 +40,302 @@ import (
+	// action, or if non-zero for the "update" the memory limit, then the new
+	// value to be set.
+	MemoryLimit quantity.Size
+
+	// ParentName is the name of the parent for the quota group if it is being
+	// created. Eventually this could be used with the "update" action to
+	// support moving quota groups from one parent to another, but that is
+	// currently not supported.
+	ParentName string
+}
+
+func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+
+	meter := snapstate.NewTaskProgressAdapterUnlocked(t)
+
+	qcs := []QuotaControlAction{}
+	err := t.Get("quota-control-actions", &qcs)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot get quota-control-action: %v", err)
+	}
+
+	// TODO: support more than one action
+	switch {
+	case len(qcs) > 1:
+		return fmt.Errorf("multiple quota group actions not supported yet")
+	case len(qcs) == 0:
+		return fmt.Errorf("internal error: no quota group actions for quota-control task")
+	}
+
+	qc := qcs[0]
+
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return err
+	}
+
+	args := quotaControlArgs{
+		St:          st,
+		AllGrps:     allGrps,
+		Action:      &qc,
+		Meter:       meter,
+		PerfTimings: perfTimings,
+	}
+
+	switch qc.Action {
+	case "create":
+		err = doCreateQuota(args)
+	case "remove":
+		err = doRemoveQuota(args)
+	case "update":
+		err = doUpdateQuota(args)
+	default:
+		err = fmt.Errorf("unknown action %q requested", qc.Action)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	t.SetStatus(state.DoneStatus)
+
+	return nil
+}
+
+func doCreateQuota(args quotaControlArgs) error {
+	st := args.St
+	action := args.Action
+	allGrps := args.AllGrps
+
+	// make sure the group does not exist yet
+	if _, ok := allGrps[action.QuotaName]; ok {
+		return fmt.Errorf("group %q already exists", action.QuotaName)
+	}
+
+	// make sure the specified snaps exist and aren't currently in another group
+	if err := validateSnapForAddingToGroup(st, action.AddSnaps, action.QuotaName, allGrps); err != nil {
+		return err
+	}
+
+	// make sure the memory limit is not zero
+	if action.MemoryLimit == 0 {
+		return fmt.Errorf("internal error, MemoryLimit option is mandatory for create action")
+	}
+
+	grp, allGrps, err := createQuotaImpl(st, action, allGrps)
+	if err != nil {
+		return err
+	}
+
+	// ensure the snap services with the group
+	opts := &ensureSnapServicesForGroupOptions{
+		allGrps: allGrps,
+	}
+	return ensureSnapServicesForGroup(st, grp, opts, args.Meter, args.PerfTimings)
+}
+
+func createQuotaImpl(st *state.State, action *QuotaControlAction, allGrps map[string]*quota.Group) (*quota.Group, map[string]*quota.Group, error) {
+	// make sure that the parent group exists if we are creating a sub-group
+	var grp *quota.Group
+	var err error
+	updatedGrps := []*quota.Group{}
+	if action.ParentName != "" {
```

Comment:
I was trying to not have 100% coverage of the cases here because it really adds up in terms of unit tests and then the PR gets to be more than 1000 lines so I just have here minimal, simple happy paths, then in the very next followup switch over the main quota_control_test.go tests testing the handlers directly in which case we then get full coverage again. I didn't want to switch those all over in this PR again because it gets bigger if I do that. Please let me know if you have an idea for how to do all the things you've asked for in this PR without creating a very large PR... I was trying to follow your recommendation of doing the simple thing first...

Reply:
Sorry the description didn't say anything about missing tests. One option would be to temporarily rewrite the old sync helpers in terms of these new ones, so use doCreateQuota from CreateQuota etc

Reply:
I suppose I could do that, let me see how that looks

Reply:
This didn't end up being too much work, so I removed the direct handler tests and instead made the exported methods call the handlers directly so all the unit tests for the exported methods are now testing the handlers